pkg/reconciler/workspaceindex/orchestrator.go

```diff
@@ -0,0 +1,506 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workspaceindex
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1/helper"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	kcpexternalversions "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	tenancyinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+	tenancylister "github.com/kcp-dev/kcp/pkg/client/listers/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/third_party/conditions/util/conditions"
+)
+
+const (
+	rootShard         = "root"
+	currentShardIndex = "shard"
+	controllerName    = "workspaceindex-orchestrator"
+)
+
+const resyncPeriod = 10 * time.Hour
+
+func NewOrchestrator(
+	kubeClient kubernetes.ClusterInterface,
+	workspaceInformer tenancyinformer.WorkspaceInformer,
+	workspaceShardInformer tenancyinformer.WorkspaceShardInformer,
+) (*Orchestrator, error) {
+	orgQueue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	workspaceQueue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+
+	c := &Orchestrator{
+		orgQueue:              orgQueue,
+		workspaceQueue:        workspaceQueue,
+		activeWorkersLock:     sync.RWMutex{},
+		kubeClient:            kubeClient,
+		workspaceIndexer:      workspaceInformer.Informer().GetIndexer(),
+		workspaceLister:       workspaceInformer.Lister(),
+		workspaceShardIndexer: workspaceShardInformer.Informer().GetIndexer(),
+		workspaceShardLister:  workspaceShardInformer.Lister(),
+		syncChecks: []cache.InformerSynced{
+			workspaceInformer.Informer().HasSynced,
+			workspaceShardInformer.Informer().HasSynced,
+		},
+		informerLock:     sync.RWMutex{},
+		shardInformers:   map[string]organizationInformer{},
+		mappingLock:      sync.RWMutex{},
+		workspaceMapping: map[string]map[string][]parsedShardStatus{},
+	}
+
+	workspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueOrg(obj)
+			c.enqueueWorkspace(rootShard, obj)
```

Comment:
usually there is a type-cast here. You know the type. Would make the whole code easier to reason about.

Reply:
Why do we need to cast? We just need to get the key for it